lib/datadog/tracing/trace_segment.rb

```diff
@@ -49,27 +63,35 @@ def initialize(
         runtime_id: nil,
         sample_rate: nil,
         sampling_priority: nil,
-        service: nil
+        service: nil,
+        tags: nil,
+        metrics: nil
       )
         @id = id
         @root_span_id = root_span_id
         @spans = spans || []
-        @tags = {}
-
-        # Set well-known tags
-        self.agent_sample_rate = agent_sample_rate
-        self.hostname = hostname
-        self.lang = lang
-        self.name = (name.frozen? ? name : name.dup)
-        self.origin = (origin.frozen? ? origin : origin.dup)
-        self.process_id = process_id
-        self.rate_limiter_rate = rate_limiter_rate
-        self.resource = (resource.frozen? ? resource : resource.dup)
-        self.rule_sample_rate = rule_sample_rate
-        self.runtime_id = runtime_id
-        self.sample_rate = sample_rate
-        self.sampling_priority = sampling_priority
-        self.service = (service.frozen? ? service : service.dup)
+
+        # Does not make an effort to move metrics out of tags
+        # The caller is expected to have done that
+        @meta = tags || {}
+        @metrics = metrics || {}
+
+        # Set well-known tags, defaulting to getting the values from tags
+        @agent_sample_rate = agent_sample_rate || agent_sample_rate_tag
+        @hostname = hostname || hostname_tag
+        @lang = lang || lang_tag
+        @name = (name || name_tag).tap { |v| break (v.frozen? ? v : v.dup) }
+        @origin = (origin || origin_tag).tap { |v| break (v.frozen? ? v : v.dup) }
+        @process_id = process_id || process_id_tag
+        @rate_limiter_rate = rate_limiter_rate || rate_limiter_rate_tag
+        @resource = (resource || resource_tag).tap { |v| break (v.frozen? ? v : v.dup) }
+        @rule_sample_rate = rule_sample_rate_tag || rule_sample_rate
+        @runtime_id = runtime_id || runtime_id_tag
+        @sample_rate = sample_rate || sample_rate_tag
+        @sampling_priority = sampling_priority || sampling_priority_tag
+        @service = (service || service_tag).tap { |v| break (v.frozen? ? v : v.dup) }
+
+        clear_known_tags
```

Comment:
Is this necessary? My take on this is the goal is to not duplicate data, but if the data already exists I don't think it matters too much.

We could instead rewrite something like `#runtime_id` to be `@runtime_id || runtime_id_tag`. Might simplify a lot of this code.

Reply:
Removed `clear_known_tags`: `TraceSegment` is immutable post initialization, and these values are unused and not publicly accessible anyway.